### PR TITLE
Bug 1476055 - Add telemetry for JEXL and other targeting errors

### PR DESCRIPTION
**Pull request by k88hudson**
This patch adds telemetry for errors that occur in JEXL string or elsewhere in the targeting environment.

If an error is thrown when a jexl string is being parsed/evaluated, a ping in the following format will be sent:

```json
{
  "client_id": "n/a",
  "action": "asrouter_undesired_event",
  "addon_version": "20180710100040",
  "impression_id": "{005deed0-e3e4-4c02-a041-17405fd703f6}",
  "locale": "en-US",
  "message_id": "some_message_id",
  "event": ["MALFORMED_EXPRESSION" | "OTHER_TARGETING_ERROR"],
}
```

To test this, you can:

1. Set the following prefs in about:config

- `browser.newtabpage.activity-stream.asrouter.snippetsUrl` to `https://gist.githubusercontent.com/k88hudson/bab3cf6e9035d5d0392311ca928480fd/raw`
- `browser.newtabpage.activity-stream.asrouterExperimentEnabled` to `true`
- `browser.ping-centre.log` to `true`
- `browser.newtabpage.activity-stream.telemetry` to `true`

2. Open the new tab a few times until you see an error come up the console (you should see `Invalid expression token: =`)
3. Ensure a `asrouter_undesired_event` event is logged

### CODE REVIEW

File: docs/v2-system-addon/data_events.md [hunk 1]
```diff
@@ -539,3 +539,20 @@ This reports the user's interaction with Activity Stream Router.
...
 }
 ```
+
+### Targeting error pings
+
+This reports when an error has occurred when parsing/evaluating a JEXL targeting string in a message.
+
+```js
+{
+  "client_id": "n/a",
```
Comment (k88hudson) [inline docs/v2-system-addon/data_events.md]: Does this look OK @ncloudioj ?

File: docs/v2-system-addon/data_events.md [hunk 2]
```diff
@@ -539,3 +539,20 @@ This reports the user's interaction with Activity Stream Router.
...
+  "action": "asrouter_undesired_event",
```
Comment (k88hudson) [inline docs/v2-system-addon/data_events.md]: Does this look OK @ncloudioj ? Or should I use some kind of asrouter user event action?

**Comment by k88hudson** [comment]
Adding @ncloudioj for review on the ping, I guess we will also need data review for this?

File: docs/v2-system-addon/data_events.md (hunk 2 above)
Comment (ncloudioj) [reply docs/v2-system-addon/data_events.md]: Yep, this is fine. Or you can make it even more specific, such as `onboarding_undesired_event`.

**Comment by ncloudioj** [comment]
LGTM! @k88hudson

Is this related to any AS Router experiment? If not, we can file a single data review request for all the metrics in AS Router.

File: lib/ASRouter.jsm [hunk 3]
```diff
@@ -268,19 +273,28 @@ class _ASRouter {
     this.messageChannel.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "ADMIN_SET_STATE", data: state});
   }
 
-  async _findMessage(messages, target, data = {}) {
-    let message;
+  _handleTargetingError(type, error, message) {
+    Cu.reportError(error);
+    if (this.dispatchToAS) {
+      this.dispatchToAS({
```
Comment (piatra) [inline lib/ASRouter.jsm]: should this (or is this) routed to the ASRouter ping center? 

https://github.com/mozilla/activity-stream/blob/868db6f54c2ac36a35fee0ad5880d0a5bbdd6d2f/lib/TelemetryFeed.jsm#L137-L141

Comment (ncloudioj) [reply lib/ASRouter.jsm]: @k88hudson any reason for not using the `ASRouterUserEvent` defined in Actions.jsm?

File: lib/ASRouterTargeting.jsm [hunk 4]
```diff
@@ -141,7 +143,12 @@ const TargetingGetters = {
 this.ASRouterTargeting = {
   Environment: TargetingGetters,
 
-  isMatch(filterExpression, target, context = this.Environment) {
+  ERROR_TYPES: {
+    MALFORMED_EXPRESSION: "MALFORMED_EXPRESSION",
+    OTHER_ERROR: "OTHER_ERROR"
+  },
+
+  isMatch(filterExpression, context = this.Environment) {
```
Comment (sarracini) [inline lib/ASRouterTargeting.jsm]: So we don't need the `target` param anymore? I guess it wasn't being used... why did we have it in the first place?

Comment (piatra) [reply lib/ASRouterTargeting.jsm]: https://github.com/mozilla/activity-stream/commit/1a74d11862da63152216a215f8f8bf21349fb73c#diff-60f0c109d5c368ec6a6b11e2f311b3ef

**Pull request merged**